### PR TITLE
Majin Buu as Render Engine

### PR DESCRIPTION
**Pull request by WebReflection**
I have done various tests and created [a new demo](https://www.youtube.com/watch?v=KZAfHQ3hnes&feature=youtu.be) that shows and proves reason behind this choice.

There is no algorithm whatsoever in hyperHTML when it comes to list of wires or DOM elements.

Until now, it has always been about unpredictable update performance if the list was similar up to a certain index or, in some case, considered fully different if just the first item was not.

This doesn't scale with any sort of big layout, infinite scrolling, pull to refresh, and you can name it story.

The goal of Majin Buu is to put an end to this madness.

It's so fast and efficient, that diverging from `any` content to `virtual` content doesn't make sense anymore: everything is virtual and handled as list without actually penalizing performance at all.

However, there are O(x*y) use cases that might be too big to compute, which is why there is now an exposed flag called `hyperHTML.MAX_LIST_SIZE` where if the square root of the Levenshtein grid is bigger than such value, the DOM will simply cleanup the list and append all new nodes.

However, this is fully handled now by `.splice` operations, fully optimized by Majin Buu and invoked through its _aura_.

I'll leave it here for a couple of days in case somebody would like to try and/or raise questions, but I'm looking forward to have this Majin Buu release of hyperHTML which will also change the way viperHTML produces output, simplifying for the future and making absolutely reliable any adoption from the backend.

Thanks for thoughts, if any.

### CODE REVIEW

**Comment by WebReflection** [comment]
this would fix https://github.com/WebReflection/hyperHTML/issues/98 too

**Comment by WebReflection** [comment]
majin_buu

**Comment by pinguxx** [comment]
So this is virtual DOM now? maybe I'm misunderstanding

**Comment by WebReflection** [comment]
@pinguxx it isn't, in the sense that hyperHTML doesn't retain a virtual copy of the DOM in memory, which is what VirtualDOM is about.

Majin Buu is a simple Levenshtein distance calculator, which is not Virtual DOM, with the ability to retrieve all optimal needed steps to transform List1 to List2.

Being JavaScript reference a no-brainer for memory, meaning that if you hold the same thing in various array the cost is **not** `thing * holders` but rather `thing` itself, since _holders_ have a memory address and not a copy, having an array that reference a list of nodes, which is the exact same list of `childNodes` the DOM retains in the view anyway, result in a zero memory overhead pattern.

The Levenshtein distance is calculated at close to C performance once, and the amount of operations per DOM node to `removeChild` or `insertBefore` is performed.

Yet, not a single abstract, virtual, representation of the DOM is held in memory.

**Comment by pinguxx** [comment]
got it, awesome stuff

File: test/infinite.html [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+          count++;
+          i++;
+        }
+        render`${newList}`;
+        view.scrollTop = scrollTop % sampleHeight;
+      }
+    }, {once: true});
+    </script>
+    <style>
+      htm, body {
```
Comment (joshgillies) [inline test/infinite.html]: htm?

Comment (WebReflection) [reply test/infinite.html]: a typo in a comment and an error in CSS ... I feel like I'm good at JS :rofl: 

Thanks, will push new stuff tomorrow !

File: test/ie/test/test.js [hunk 2]
```diff
@@ -334,14 +341,15 @@ tressa.async(function (done) {
   tressa.assert(div.firstChild.firstChild === br, 'one child is added');
   hyperHTML.bind(div)(_templateObject14, hr);
   tressa.assert(div.firstChild.firstChild === hr, 'one child is changed');
+  debugger;
```
Comment (GBarthos) [inline test/ie/test/test.js]: left a `debugger` on purpose ?

Comment (WebReflection) [reply test/ie/test/test.js]: good catch

**Pull request merged**